### PR TITLE
Add support for textDocument/prepareRename (added in LSP 3.12)

### PR DESCRIPTION
**Pull request by thomasjm**
Cf. issue #134

### CODE REVIEW

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs [hunk 1]
```diff
@@ -2586,6 +2611,37 @@ deriveJSON lspOptions ''RenameParams
...
+            valid at the given position.
+    error: code and message set in case an exception happens during the
+           prepare rename request.
+
+-}
+
+-- {\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"textDocument/rename\",\"params\":{\"textDocument\":{\"uri\":\"file:///home/alanz/mysrc/github/alanz/haskell-lsp/src/HieVscode.hs\"},\"position\":{\"line\":37,\"character\":17},\"newName\":\"getArgs'\"}}
+
+type PrepareRenameRequest  = RequestMessage ClientMethod TextDocumentPositionParams Range
+type PrepareRenameResponse = ResponseMessage Range
```
Comment (bubba) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Should we handle the case where it can return a range and a placeholder?

**Comment by bubba** [comment]
Can you update the helper functions inside Capabilities.hs to reflect that this is available in 3.12?

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs (hunk 1 above)
Comment (thomasjm) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Yes, I'll add a type for that

**Comment by thomasjm** [comment]
Ah I see -- please double check it looks good now.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs [hunk 2]
```diff
@@ -307,6 +307,31 @@ data DocumentLinkOptions =
...
+         */
+        prepareProvider?: boolean;
+}
+-}
+
+data RenameOptions =
+  RenameOptionsStatic Bool
+  | RenameOptions
+    { -- |Renames should be checked and tested before being executed.
+      _prepareProvider :: Maybe Bool
```
Comment (lorenzo) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Would you mind using this other comment notation?

```haskell
data RenameOptions =
	  RenameOptionsStatic Bool
	  | RenameOptions
	    {  _prepareProvider :: Maybe Bool -- ^ Renames should be checked and tested before being executed.
```

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs (hunk 1 above)
Comment (thomasjm) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Just added the type salad for this

**Pull request merged**